### PR TITLE
Variable CSA response curve

### LINKED ISSUE #17: Variable CSA response curve
@simonspa I'm working on writing a simulation for the MuPix10 with Allpix-Squared. We have a CSA, but the impulse response curve doesn't look like the one from the CSADigitizer module, it falls linearly. A possible parameterization is `A * (1-exp(-x/tauR_) ) - tauLinearF * x`, and `0` when when the linear part becomes larger than the exponential one.

My question is what the best way is to implement this, if we want to upstream it. My idea was to add a new model (so that there would be `simple`, `csa` and `mupix`). However, this would require a model-dependent impulse response here:
https://github.com/allpix-squared/allpix-squared/blob/8e36b5156dd16f4db5d64751f91109e815ff5bb5/src/modules/CSADigitizer/CSADigitizerModule.cpp#L189-L197

Since it's only called once I don't see a big problem there. Let me know what you think. Also, the documentation for the module would require some rewriting.

### PR DESCRIPTION
**Pull request by stephanlachnit**
Closes #17

@simonspa so this would be the my initial idea for the implementation. It's not polished by any means, just a concept.

I thought figured offloading everything to a different files makes changes or additional implementations in the future easier. Especially with the impulse response being part of a class, any type of response can be implemented. In theory, one could even move the entire pulse amplification loop to the class, to implement features like saturation.

### CODE REVIEW

**Comment by stephanlachnit** [comment]
@simonspa this is now how I would implement it. I tested the `simple` type and I works as expected. I still need to check some corner cases for the `tf1` type, but in principle everything should be there.

I would add a CI test and update the documentation once you're happy with the implementation.

**Comment by stephanlachnit** [comment]
Btw the github CI needs to be updated for the latest commits:

```
|16:19:25.553|   (FATAL) Error in the configuration:
                         Key 'workers' in global section does not exist
                         The configuration needs to be updated. Cannot continue.
```

**Comment by stephanlachnit** [comment]
Alright I tested if the function is invalid and if the amount of parameters given isn't the same as the function takes, both throw as expected. In the first case, I found no way to disable the output of the root error message, but that isn't necessarily bad as it helps to understand what's wrong in the formula. It isn't pretty though.

One open question is the default value. Right now I just use some formula with random values that roughly imitate what the MuPix10 does. However, I think it's rather ugly. How does simply returning 0 sound to you?

**Comment by schuetzepaul** [comment]
> Btw the github CI needs to be updated for the latest commits:
> 
> ```
> |16:19:25.553|   (FATAL) Error in the configuration:
>                          Key 'workers' in global section does not exist
>                          The configuration needs to be updated. Cannot continue.
> ```

That's an interesting behaviour... I will have a look at this, thanks!

**Comment by simonspa** [comment]
Hej @stephanlachnit 

thanks for the ping - this approach to me looks very elegant and clean, you're using the same code for all models. I have two (very nit-picking) ideas:
* How about making the `TF1*` a `std::unique_ptr<TF1>`? Just for the sake of not having raw pointers flying around. Not that it would change something.
* How about naming the new parameters `response_function` and `response_parameters`? Not sure about this, but it makes the relation a bit more clear? @schuetzepaul opinion?
* I'm thinking about the `model = "tf1"`, maybe we could come up with a more "speaking" parameter name? For the internal type it's fine though...

Apart from that: :rocket: 

* If you'd add some unit tests and the documentation, I'd say this is ready to go.
* I would simply *not* set a default as response function. This basically forces the user to provide it when selecting `model = TF1`.
* Concerning the Mupix10 response - how about adding this as one of the examples to the `README`'s last section?
* For now, don't worry about the CI, there seems to be a more fundamental issue somewhere... :)

Thanks a lot, nice piece of code! :)

Simon

**Comment by stephanlachnit** [comment]
> How about making the TF1* a std::unique_ptr<TF1>? Just for the sake of not having raw pointers flying around. Not that it would change something.

Oh yeah good idea.

> How about naming the new parameters response_function and response_parameters? Not sure about this, but it makes the relation a bit more clear?

Agreed.

> I'm thinking about the model = "tf1", maybe we could come up with a more "speaking" parameter name? For the internal type it's fine though...

I don't like it either, but I don't a fitting name for it. `variable` maybe? At least when combined with "variable model" it makes more sense then for example `function`. Or `custom`, that would also work.

> If you'd add some unit tests and the documentation, I'd say this is ready to go.

Alright then I will start working on that.

> Concerning the Mupix10 response - how about adding this as one of the examples to the README's last section?

Good idea.

**Comment by simonspa** [comment]
> I don't like it either, but I don't a fitting name for it. `variable` maybe? At least when combined with "variable model" it makes more sense then for example `function`. Or `custom`, that would also work.

`custom` sounds great to me!

**Comment by stephanlachnit** [comment]
Ok I think I'm almost done. Two things left:

- Feedback on the documentation. I'm not the best with words.
- I don't quite understand the comment at the bottom of the unit tests. How should I pick them? Are the unit tests just supposed to pass with no throwing error or can we check internal values?

File: src/modules/CSADigitizer/README.md [hunk 1]
```diff
@@ -32,16 +33,22 @@ If this behavior is not desired, the `ignore_polarity` parameter can be set to c
...
+* `feedback_capacitance` : The feedback capacity to the amplifier circuit. Defaults to 5e-15 F.
 * `krummenacher_current` : The feedback current setting of the CSA. Defaults to 20 nA.
 * `detector_capacitance` : The detector capacitance. Defaults to 100 e-15 F.
 * `amp_output_capacitance` : The capacitance at the amplifier output. Defaults to 20 e-15 F.
 * `transconductance` : The transconductance of the CSA feedback circuit. Defaults to 50e-6 C/s/V.
 * `temperature` : Defaults to 293.15K.
 
+#### Parameters for the custom model
+* `response_function` : A 1-dimensional [`ROOT::TFormula`](https://root.cern.ch/doc/master/classTFormula.html) expression for the impulse response function.
+* `response_parameters` : Array of the parameters in the repsonse function.
```
Comment (simonspa) [inline src/modules/CSADigitizer/README.md:50]: Typo `repsonse`.

Maybe add that the number of parameters needs to match the one given in the formula?

File: src/modules/CSADigitizer/README.md [hunk 2]
```diff
@@ -14,6 +14,8 @@ The impulse response function of this transfer function is convoluted with the c
 This module can be steered by either providing all contributions to the transfer function as parameters within the `csa` model, or using a simplified parametrization providing rise time and feedback time.
 In the latter case, the parameters are used to derive the contributions to the transfer function (see e.g. [@binkley] for calculation of transconductance).
 
+Alternatively a custom impulse response functions can be implented in the `custom` model.
```
Comment (simonspa) [inline src/modules/CSADigitizer/README.md:17]: maybe "...can be provided by using the `custom` model."?

File: src/modules/CSADigitizer/README.md [hunk 3]
```diff
@@ -82,5 +89,18 @@ clock_bin_tot = 25.0ns
 
 
 
+```ini
```
Comment (simonspa) [inline src/modules/CSADigitizer/README.md:92]: Maybe add a comment above what this one is for (i.e. MuPix10)? And remove the excessive newlines after.

**Comment by simonspa** [comment]
Minor detail: Feel free to pile up several commits, that makes them more atomic and in the end eases finding a problem later on e.g. using `git bisect`. Otherwise it just points to the large blob of code that adds a com0pletely new feature.

**Comment by simonspa** [comment]
Concerning the test cases:

these current are rather system tests than unit tests, meaning we have no introspection into the code but simply parse the output of a single-event run with the relevant modules active. This approach is described here in the manual:
https://project-allpix-squared.web.cern.ch/project-allpix-squared/usermanual/allpix-manualch10.html#x11-22700010.6

The `PASS` condition is picked up automatically by CTest and is compared as regex against the stdout of the simulation defined by the test config. There is an additional `PASSOSX` because unfortunately STL does not guarantee implementations of random distributions, but only of the underlying generators. This means we get divergent results on different platforms. At some point in the future we might want to ship our own distribution implementations to solve this... :)

**Comment by stephanlachnit** [comment]
Ok thanks for the explanation for the unit tests. I now use

```ini
#PASS [R:CSADigitizer:mydetector] Pixel (2,0): time 7clk, signal 515clk
#PASSOSX [R:CSADigitizer:mydetector] Pixel (2,0): time 7clk, signal 515clk
```

since this will check if the function is actually applied. I could also use

```ini
#PASS [C:CSADigitizer:mydetector] Response function successfully initialized with 3 parameters
#PASSOSX [C:CSADigitizer:mydetector] Response function successfully initialized with 3 parameters
```

which describes a bit more what is actually tested, but is a bit weaker.

**Comment by stephanlachnit** [comment]
Also regarding the documentation, in https://github.com/allpix-squared/allpix-squared/blob/master/doc/usermanual/chapters/testing.tex a lot of tests are missing so I didn't add the new one, should I add it?

**Comment by simonspa** [comment]
> Ok thanks for the explanation for the unit tests. I now use
> 
> ```ini
> #PASS [R:CSADigitizer:mydetector] Pixel (2,0): time 7clk, signal 515clk
> #PASSOSX [R:CSADigitizer:mydetector] Pixel (2,0): time 7clk, signal 515clk
> ```
> 
> since this will check if the function is actually applied. I could also use
> 
> ```ini
> #PASS [C:CSADigitizer:mydetector] Response function successfully initialized with 3 parameters
> #PASSOSX [C:CSADigitizer:mydetector] Response function successfully initialized with 3 parameters
> ```
> 
> which describes a bit more what is actually tested, but is a bit weaker.

You could also just add two tests :)

**Comment by simonspa** [comment]
> Also regarding the documentation, in https://github.com/allpix-squared/allpix-squared/blob/master/doc/usermanual/chapters/testing.tex a lot of tests are missing so I didn't add the new one, should I add it?

I still hope to get that autogenerated from CMake at some point: https://gitlab.cern.ch/allpix-squared/allpix-squared/-/issues/207 - so from my point of view: don't bother. :)

**Comment by stephanlachnit** [comment]
> You could also just add two tests :)

Yup, just did. Note though: maybe add a trailing zero in the second number of the tests like with the first number (e.g. `test_06-09`), right now ctest doesn't execute them in the natural order.

I think I fixed everything you pointed out. I will probably come back to the examples and exchange some values for the MuPix10 when I have better ones.

**Comment by stephanlachnit** [comment]
Ah found typos. Fixed now.

**Comment by simonspa** [comment]
Awesome, thanks a lot, @stephanlachnit for all your work and this very neat feature!

Final CI run on Gitlab here: https://gitlab.cern.ch/allpix-squared/allpix-squared/-/merge_requests/455
and then I'm merging this.

**Comment by simonspa** [comment]
> > You could also just add two tests :)
> 
> Yup, just did. Note though: maybe add a trailing zero in the second number of the tests like with the first number (e.g. `test_06-09`), right now ctest doesn't execute them in the natural order.
> 
> I think I fixed everything you pointed out. I will probably come back to the examples and exchange some values for the MuPix10 when I have better ones.

Tests are executed in parallel in the CI, so there is no natural order in running them. But I agree, it's not very nice for ordering. We are planning to make the tests more atomic and eventually move them to their respective modules - then the ordering problem would be largely gone anyway.

**Comment by simonspa** [comment]
:clap:
